x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleTask.java

```diff
@@ -204,37 +238,144 @@ public static String exceptionToString(Exception ex) {
+                    snapshotId.getName(),
+                    policyName
+                );
+                return currentState;
+            }
+
+            Map<String, SnapshotLifecyclePolicyMetadata> snapLifecycles = new HashMap<>(snapMeta.getSnapshotConfigurations());
+            SnapshotLifecyclePolicyMetadata policyMetadata = snapLifecycles.get(policyName);
+            if (policyMetadata == null) {
+                logger.warn("failed to pre-register snapshot [{}] in policy [{}]: policy not found", snapshotId.getName(), policyName);
+                return currentState;
+            }
+
+            Set<SnapshotId> preRegisteredSnapshots = policyMetadata.getPreRegisteredSnapshots();
+            Set<SnapshotId> runningSnapshots = currentlyRunningSnapshots(policyName, currentState);
+            SnapshotLifecyclePolicyMetadata.Builder newPolicyMetadata = SnapshotLifecyclePolicyMetadata.builder(policyMetadata);
+            SnapshotLifecycleStats stats = snapMeta.getStats();
+            long unrecordedFailures = 0;
+            Set<SnapshotId> newPreRegisteredSnapshots = new HashSet<>();
+            for (SnapshotId snapshot : preRegisteredSnapshots) {
+                if (runningSnapshots.contains(snapshot)) {
+                    newPreRegisteredSnapshots.add(snapshot);
+                } else {
+                    stats.snapshotFailed(policyName);
+                    unrecordedFailures++;
+                    newPolicyMetadata.setLastFailure(
+                        new SnapshotInvocationRecord(
+                            snapshot.getName(),
+                            null,
+                            Instant.now().toEpochMilli(),
+                            String.format(
+                                Locale.ROOT,
+                                "found pre-registered snapshot [%s] which is no longer running, assuming failed.",
+                                snapshot.getName()
+                            )
+                        )
+                    );
+                }
+            }
+            newPreRegisteredSnapshots.add(snapshotId);
+            newPolicyMetadata.setPreRegisteredSnapshots(newPreRegisteredSnapshots);
+            newPolicyMetadata.setInvocationsSinceLastSuccess(policyMetadata.getInvocationsSinceLastSuccess() + unrecordedFailures);
+
+            snapLifecycles.put(policyName, newPolicyMetadata.build());
+            SnapshotLifecycleMetadata lifecycleMetadata = new SnapshotLifecycleMetadata(
+                snapLifecycles,
+                currentSLMMode(currentState),
+                stats
+            );
+            Metadata currentMeta = currentState.metadata();
+            return ClusterState.builder(currentState)
+                .metadata(Metadata.builder(currentMeta).putCustom(SnapshotLifecycleMetadata.TYPE, lifecycleMetadata))
+                .build();
+        }
+
+        @Override
+        public void onFailure(Exception e) {
```

Comment:
Do you think there is a better way to handle this failure? This will be called if there is a failure while attempting to preRegister a snapshot.